src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlQueueClient.cs

```diff
@@ -316,5 +335,71 @@ public async Task<bool> KeepAliveJobAsync(JobInfo jobInfo, CancellationToken can
                 throw;
             }
         }
+
+        public async Task ArchiveJobsAsync(byte queueType, CancellationToken cancellationToken)
+        {
+            using SqlConnectionWrapper conn = await _sqlConnectionWrapperFactory.ObtainSqlConnectionWrapperAsync(cancellationToken, false);
+            using SqlCommandWrapper cmd = conn.CreateRetrySqlCommand();
+            VLatest.ArchiveJobs.PopulateCommand(cmd, queueType);
+            await cmd.ExecuteNonQueryAsync(cancellationToken);
+        }
+
+        public async Task ExecuteWithHeartbeatAsync(byte queueType, long jobId, long version, Func<CancellationToken, Task> action, TimeSpan heartbeatPeriod, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNull(action, nameof(action));
+
+            using var tokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+            using var timer = new PeriodicTimer(heartbeatPeriod);
+
+            CancellationToken timerToken = tokenSource.Token;
+
+            Task heartBeatTask = HeartbeatLoopAsync(queueType, jobId, version, timer, timerToken);
+            Task<Task> actionTask = action.Invoke(timerToken).ContinueWith(
+                _ =>
+                {
+                    tokenSource.Cancel();
+                    return Task.CompletedTask;
+                },
+                TaskScheduler.Current);
+
+            try
+            {
+                await Task.WhenAll(actionTask, heartBeatTask);
```

Comment:
>await Task.WhenAll(actionTask, heartBeatTask); [](http://example.com/codeflow?start=16&length=46)

@brendankowitz Currently it waits on heartbeat task and checks its state. This is different from my previous implementation and is not desired, as heartbeats are meant to be fire and forget. How would you suggest to fix this?

Reply:
L360 cancels the timer's token as soon as the action is done, this should be immediate. However, it should still be ok to just wait for action.

```suggestion
                await actionTask;
```

Reply:
I replaced all code by original implementation based on Timer. I like its conciseness.